docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,822 @@
+stencils, unless an origin and domain are explicitly specified. Further, the domain can be accessed
+in the storage through the :code:`domain` attribute of the storage. The result is again a storage
+with no halo, such that the index :code:`storage.domain[0, 0, 0]` refers to a corner of the inner
+domain denoted by the halo.
+
+The halo can be changed after allocation by assigning to :code:`storage.halo` using the same values
+as in the :code:`halo` parameter above.
+
+The halo passed when allocating is used as a hint for alignment. If no :code:`aligned_index` is
+explicitly specified, the lower indices of the :code:`halo` parameter are used to align the memory
+which, when coinciding with the origin when calling stencils, potentially has performance benefits.
+When changing the halo after allocation, this has however no impact on the alignment and it is in
+general best to specify the same halo that is used in the computations already when allocating.
+
+Storage Attributes and NumPy API functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As we mentioned above, keeping compatibility with NumPy is an essential requirement for GT4Py
+Storages and therefore they should support the most relevant parts of the NumPy API. An
+initial proposal of supported features is presented here. By features we mean NumPy functions
+(:code:`np.function()` -like) that work well with GT4Py storages, as well as attributes
+(:code:`ndarray.attribute`) and methods (:code:`ndarray.method()`) of the :code:`ndarray` class.
+
+We are aware that this is by no means an exhaustive list and we ask for additional input from the
+community to collect other features that should be supported.
+
+NumPy Functions
+===============
+:code:`np.all`
+    same semantics as :code:`np.logical_and.reduce`, when applied to all axes.
+
+:code:`np.any`
+    same semantics as :code:`np.logical_or.reduce`, when applied to all axes.
+
+:code:`np.max`
+    same semantics as :code:`np.max.reduce`, when applied to all axes.
+
+:code:`np.min`
+    same semantics as :code:`np.min.reduce`, when applied to all axes.
+
+:code:`np.transpose`
+    permutation of the axes. In addition to the parameters of :code:`np.transpose`, when applied to
+    :code:`ndarray`'s, :code:`axes` can be the usual strings to represent the :code:`axes` attribute
+    of the resulting storage. See also the :code:`reinterpret` method below.
+
+
+Attributes and Properties
+=========================
+:code:`Storage` s have the following attributes:
+
+:code:`__array_interface__: Dict[str, Any]`
+    The *Array Interface* descriptor of this storage (only supported on instances with an
+    actual host buffer).
+
+:code:`__cuda_array_interface__: Dict[str, Any]`
+    The *CUDA Array Interface* descriptor of this storage (only supported on instances with an
+    actual GPU device buffer).
+
+:code:`alignment: int`
+    Alignment size value given at creation time
```

Comment:
So, a given point in the grid is aligned to a certain `alignment` size. (we could change the name, I see how it can be confusing)

So, if the memory address of a grid point is `addr`, then it is considered aligned if `addr%alignment==0`. This is a performance feature. If you always go through the `defaults` parameter and specify your `halo` as the value that is actually used in the computation, you never have to worry about this.

Otherwise, you may or may not suffer a performance penalty.

Reply:
The value for `alignment` which was given at creation time.

Agreed, I'll change it to re-explain the attributes in general.